### PR TITLE
Convert NetCDF Char Arrays to String Arrays for IODA v2 Compliance  

### PR DESCRIPTION
**Pull request by amstokely**
### Description:  
This PR updates NetCDF variable handling by converting character array variables to string array variables, ensuring compliance with the IODA v2 format.  

### Testing:  
- Added tests to verify that former character array variables are now written as string array variables.  
- Confirmed that the number of dimensions for these variables is 1 instead of 2.  
- Relevant test cases can be found [here](https://github.com/NCAR/obs2ioda/blob/63fd4add624980b389f0a8e09762a7f18bad21fa/test/integration/check_variables.cc#L43) and [here](https://github.com/NCAR/obs2ioda/blob/63fd4add624980b389f0a8e09762a7f18bad21fa/test/integration/check_variables.cc#L74).

### CODE REVIEW

File: obs2ioda-v2/src/define_mod.f90 [hunk 1]
```diff
@@ -159,8 +159,8 @@ module define_mod
       nf90_float, &
       nf90_int64, &
       nf90_char,  &
-      nf90_char,  &
-      nf90_char   &
+      nf90_string,  &
```
Comment (jim-p-w) [inline obs2ioda-v2/src/define_mod.f90:162]: It looks like the value of `type_var_info` is checked in various places, e.g.
https://github.com/NCAR/obs2ioda/blob/b109765ad2a72cbe0c26a6e5969121d3b0a223a5/obs2ioda-v2/src/prepbufr_mod.f90#L821

Is it possible some of these checks will need to change to accommodate the new types?

File: obs2ioda-v2/src/define_mod.f90 [hunk 2]
```diff
@@ -171,8 +171,8 @@ module define_mod
       'nlocs     ', 'null      ', &
       'nlocs     ', 'null      ', &
       'ndatetime ', 'nlocs     ', &
-      'nstring   ', 'nlocs     ', &
-      'nstring   ', 'nvars     '  &
+      'nlocs     ', 'null      ', &
```
Comment (jim-p-w) [inline obs2ioda-v2/src/define_mod.f90:174]: These changes look like the usage of `dim_var_info` might need to change:
https://github.com/NCAR/obs2ioda/blob/b109765ad2a72cbe0c26a6e5969121d3b0a223a5/obs2ioda-v2/src/ncio_mod.f90#L210
https://github.com/NCAR/obs2ioda/blob/b109765ad2a72cbe0c26a6e5969121d3b0a223a5/obs2ioda-v2/src/ncio_mod.f90#L213

File: obs2ioda-v2/src/define_mod.f90 (hunk 1 above)
Comment (amstokely) [reply obs2ioda-v2/src/define_mod.f90:162]: The tests confirm that these changes successfully write string array variables for the variables that were originally character arrays. I'll review `ncio_mod` again to ensure no updates are needed for `nc_char` logic.

Comment (jim-p-w) [reply obs2ioda-v2/src/define_mod.f90:162]: > The tests confirm that these changes successfully write string array variables for the variables that were originally character arrays. I'll review `ncio_mod` again to ensure no updates are needed for `nc_char` logic.

That may be, but are the values correct? It looks like `station_id` and `variable_names` are changing from nf90_char to nf90_string, and line 821 of prepbufr_mod.f90 checks for a type of nf90_char. Within that if clause there's a check at line 828:
`else if ( trim(name_var_info(i)) == 'station_id' ) then`
which will no longer be true.

Comment (amstokely) [reply obs2ioda-v2/src/define_mod.f90:162]: I just updated this logic to account for the station_id type change.

File: obs2ioda-v2/src/define_mod.f90 (hunk 2 above)
Comment (amstokely) [reply obs2ioda-v2/src/define_mod.f90:174]: Since `if (ufo_vars_getindex(name_ncdim, dim_var_info(2, i)) > 0)` is now false for `nf90_string` variables, those variables are added in the `else` statement. However, this `if` statement is now redundant, so I'll remove it.

File: obs2ioda-v2/src/hsd.f90 [hunk 3]
```diff
@@ -663,9 +663,11 @@ subroutine read_HSD(ccyymmddhhnn, inpdir, do_superob, superob_halfwidth)
            else if ( type_var_info(i) == nf90_char ) then
               if ( trim(name_var_info(i)) == 'datetime' ) then
                  xdata(1,1)%xinfo_char(iloc,i) = datetime
-              else if ( trim(name_var_info(i)) == 'station_id' ) then
```
Comment (jim-p-w) [inline obs2ioda-v2/src/hsd.f90:666]: There's another check for station_id in an `if .. nf90_char` clause at line 773.

Comment (amstokely) [reply obs2ioda-v2/src/hsd.f90:666]: Thanks for pointing this out...just fixed it.

File: obs2ioda-v2/src/hsd.f90 [hunk 4]
```diff
@@ -773,6 +773,10 @@ subroutine read_HSD(ccyymmddhhnn, inpdir, do_superob, superob_halfwidth)
               else if ( trim(name_var_info(i)) == 'station_id' ) then
                  xdata(1,1)%xinfo_char(iloc,i) = 'ahi_himawari8'
```
Comment (jim-p-w) [inline obs2ioda-v2/src/hsd.f90:774]: Consider removing the outdated check for station_id?

File: obs2ioda-v2/src/define_mod.f90 [hunk 5]
```diff
@@ -159,8 +159,8 @@ module define_mod
       nf90_float, &
       nf90_int64, &
       nf90_char,  &
```
Comment (st-ncar) [inline obs2ioda-v2/src/define_mod.f90:161]: I am not entirely sure what the scope of this PR is - are we attempting to remove all instances of nf90_char? If so, do we need to update this entry and how we process `datetime`? It seems to be an edge case, `datetime` is currently not written to file (see settings in `iflag_conv`).

Comment (amstokely) [reply obs2ioda-v2/src/define_mod.f90:161]: I only updated variables actively used by `obs2ioda`. As far as I know, `datetime` is officially deprecated in IODA, but for consistency, I'll change its type from `nf90_char` to `nf90_string`.

File: obs2ioda-v2/src/ncio_mod.f90 [hunk 6]
```diff
@@ -351,6 +345,7 @@ subroutine write_obs (filedate, write_opt, outdir, itim)
       end do var_info_loop
 
       if ( write_opt == write_nc_radiance .or. write_opt == write_nc_radiance_geo ) then
+         allocate(scan_position_values(xdata(ityp, itim)%nlocs, nsen_info))
```
Comment (st-ncar) [inline obs2ioda-v2/src/ncio_mod.f90:348]: This line leads to a compilation error for me. It looks like `scan_position_values` is declared as 1D array, but we try to allocate a 2D array here.

File: obs2ioda-v2/src/ncio_mod.f90 [hunk 7]
```diff
@@ -376,6 +370,7 @@ subroutine write_obs (filedate, write_opt, outdir, itim)
          end if
          deallocate (ichan)
          deallocate (obserr)
+         deallocate (scan_position_values)
```
Comment (st-ncar) [inline obs2ioda-v2/src/ncio_mod.f90:373]: This causes a runtime error. Possible duplicate of line 359?

File: obs2ioda-v2/src/radiance_mod.f90 [hunk 8]
```diff
@@ -1122,7 +1122,9 @@ subroutine sort_obs_radiance(filedate, nfgat)
          else if ( type_var_info(i) == nf90_char ) then
```
Comment (st-ncar) [inline obs2ioda-v2/src/radiance_mod.f90:1122]: Throughout the files, are there any variables of type `nf90_char` left? It seems like `datetime` should now be of type `nf90_string` as well.

Comment (amstokely) [reply obs2ioda-v2/src/radiance_mod.f90:1122]: There aren’t any. I’ll update the testing branch and run tests on these changes. It looks like a few small edge cases keep slipping through.

Comment (st-ncar) [reply obs2ioda-v2/src/radiance_mod.f90:1122]: If there are no more `nf90_char` left, I think we need to move all the relevant function calls from the `nf90_char` if blocks to the`nf90_string` blocks. Otherwise that code never gets executed.

File: obs2ioda-v2/src/satwnd_mod.f90 [hunk 9]
```diff
@@ -489,7 +489,9 @@ subroutine sort_obs_satwnd(filedate, nfgat)
             else if ( type_var_info(i) == nf90_char ) then
                if ( trim(name_var_info(i)) == 'datetime' ) then
```
Comment (st-ncar) [inline obs2ioda-v2/src/satwnd_mod.f90:490]: For example, this if condition is never satisfied, so it would seem that `xdata` is missing `datetime`.

File: obs2ioda-v2/src/hsd.f90 [hunk 10]
```diff
@@ -663,9 +663,11 @@ subroutine read_HSD(ccyymmddhhnn, inpdir, do_superob, superob_halfwidth)
            else if ( type_var_info(i) == nf90_char ) then
               if ( trim(name_var_info(i)) == 'datetime' ) then
```
Comment (st-ncar) [inline obs2ioda-v2/src/hsd.f90:664]: Another if clause that can perhaps no longer be reached?

Comment (amstokely) [reply obs2ioda-v2/src/hsd.f90:664]: Since the `datetime` variable isn't routinely used and changing it from `nf90_char` to `nf90_string` could break existing logic, I suggest keeping it as `nf90_char` for now. We can revisit this later, once the major `obs2ioda` IO refactor is complete.

File: obs2ioda-v2/src/ncio_mod.f90 (hunk 7 above)
Comment (amstokely) [reply obs2ioda-v2/src/ncio_mod.f90]: Just fixed this.

File: obs2ioda-v2/src/satwnd_mod.f90 (hunk 9 above)
Comment (amstokely) [reply obs2ioda-v2/src/satwnd_mod.f90]: I moved all code changes to `ncio_mod` to eliminate the need to update `nf90_char` dependent logic in files outside of `nf90_mod.f90`.

File: obs2ioda-v2/src/radiance_mod.f90 (hunk 8 above)
Comment (amstokely) [reply obs2ioda-v2/src/radiance_mod.f90]: See below comments.

**Review by st-ncar — APPROVED** [review]
All my comments have been addressed.

**Review by jim-p-w — APPROVED** [review]
